File: leetcode/002-Add-Two-Numbers/go/Solution.go
package _go

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	p := new(ListNode)
	dump := p
	l1_val, l2_val, c, sum := 0, 0, 0, 0

	for l1 != nil || l2 != nil {
		if l1 != nil {
			l1_val = l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			l2_val = l2.Val
			l2 = l2.Next
		}

		sum = l1_val + l2_val + c
		p.Next = new(ListNode)
		p = p.Next
		p.Val, c = sum % 10, sum / 10

		l1_val, l2_val = 0, 0
	}

	if c == 1 {
		p.Next = &ListNode{1, nil}
	}

	return dump.Next
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: leetcode/002-Add-Two-Numbers/scala/Solution.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
  def addTwoNumbers(l1: ListNode, l2: ListNode): ListNode = {
    val dump = new ListNode
    var sum, c, l1_val, l2_val:Int = 0
    var (p, pl1, pl2) = (dump, l1, l2)

    while(pl1 != null || pl2 != null) {
      if (pl1 != null) {
        l1_val = pl1.x
        pl1 = pl1.next
      }

      if (pl2 != null) {
        l2_val = pl2.x
        pl2 = pl2.next
      }

      sum = c + l1_val + l2_val
      p.next = new ListNode(sum % 10)
      p = p.next
      c = sum / 10

      l1_val = 0; l2_val = 0
    }

    if (c == 1) {
      p.next = new ListNode(1)
    }

    return dump.next
  }
}